extensions/amp-embedly/0.1/amp-embedly.js

```diff
@@ -0,0 +1,190 @@
+/**
+ * Copyright 2018 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {Services} from '../../../src/services';
+import {addParamsToUrl} from '../../../src/url';
+import {dict} from '../../../src/utils/object';
+import {getIframe} from '../../../src/3p-frame';
+import {isLayoutSizeDefined} from '../../../src/layout';
+import {user} from '../../../src/log';
+
+/** @private @const {string} */
+const TAG = 'amp-embedly';
+
+/**
+ * @const {string}
+ */
+const BASE_API_URL = 'https://api.embedly.com/1/oembed?';
+
+/**
+ * Regex used to extract src from returned oEmbed data.
+ * @const {RegExp}
+ * */
+const SRC_REGEXP = /src="([^"]+)"/;
+
+/**
+ * oEmbed resource types.
+ * @const {Object<{PHOTO: string, VIDEO: string, LINK: string, RICH: string}>}
+ */
+const resourceType = Object.freeze({
+  PHOTO: 'photo',
+  VIDEO: 'video',
+  LINK: 'link',
+  RICH: 'rich',
+});
+
+/**
+ * Implementation of the amp-embedly component.
+ * See {@link ../amp-embedly.md} for the spec.
+ */
+export class AmpEmbedly extends AMP.BaseElement {
+  /** @param {!AmpElement} element */
+  constructor(element) {
+    super(element);
+
+    /** @const @private {!../../../src/service/xhr-impl.Xhr} */
+    this.xhr_ = Services.xhrFor(this.win);
+
+    /** @private {?string}  */
+    this.key_ = null;
+
+    /** @private {?string}  */
+    this.url_ = null;
+
+    /** @private {?Element} */
+    this.iframe_ = null;
+  }
+
+  /** @override */
+  buildCallback() {
+    this.key_ = user().assert(
+        this.element.getAttribute('data-key'),
```

Comment:
Consider describing a separate element to define the data-key that all instances consume from. It would be delivered as part of the same extension.

Something like:

`<amp-embedly-key value="XXX"></amp-embedly-key>`.

On `buildCallback` [this component would register a service](https://github.com/ampproject/amphtml/blob/master/src/element-service.js) that the `AmpEmbedly` instances would consume from.

Let me know if you need help with this.

Reply:
Created EmbedlyService and moved xhr request logic there. 
`amp-embedly-key` receives key value and sets key in element service.

Updated `amp-embedly` component to consume `EmbedlyService`.

These three, two components + service are registered as part of the extension.

Created `Services.embedlyServiceForDoc` static method.